contracts/NFTStorefrontV2.cdc

```diff
@@ -0,0 +1,706 @@
+            // The first receiver should therefore either be the seller, or an agreed recipient for
+            // any unpaid cuts.
+            var residualReceiver: &{FungibleToken.Receiver}? = nil
+            // Pay the comission 
+            // Pay each beneficiary their amount of the payment.
+
+            for cut in self.details.saleCuts {
+                if let receiver = cut.receiver.borrow() {
+                   let paymentCut <- payment.withdraw(amount: cut.amount)
+                    receiver.deposit(from: <-paymentCut)
+                    if (residualReceiver == nil) {
+                        residualReceiver = receiver
+                    }
+                }
+            }
+
+            assert(residualReceiver != nil, message: "No valid payment receivers")
+
+            // At this point, if all recievers were active and availabile, then the payment Vault will have
+            // zero tokens left, and this will functionally be a no-op that consumes the empty vault
+            residualReceiver!.deposit(from: <-payment)
+
+            // If the listing is purchased, we regard it as completed here.
+            // Otherwise we regard it as completed in the destructor.
+
+            emit ListingCompleted(
+                listingResourceID: self.uuid,
+                storefrontResourceID: self.details.storefrontID,
+                purchased: self.details.purchased,
+                nftType: self.details.nftType,
+                nftUUID: self.details.nftUUID,
+                nftID: self.details.nftID,
+                salePaymentVaultType: self.details.salePaymentVaultType,
+                salePrice: self.details.salePrice,
+                customID: self.details.customID,
+                commissionAmount: self.details.commissionAmount,
+                expiry: self.details.expiry
+            )
+
+            return <-nft
+        }
+
+        // destructor
+        //
+        destroy () {
+            // If the listing has not been purchased, we regard it as completed here.
+            // Otherwise we regard it as completed in purchase().
+            // This is because we destroy the listing in Storefront.removeListing()
+            // or Storefront.cleanup() .
+            // If we change this destructor, revisit those functions.
+            if !self.details.purchased {
+                emit ListingCompleted(
+                    listingResourceID: self.uuid,
+                    storefrontResourceID: self.details.storefrontID,
+                    purchased: self.details.purchased,
+                    nftType: self.details.nftType,
+                    nftUUID: self.details.nftUUID,
+                    nftID: self.details.nftID,
+                    salePaymentVaultType: self.details.salePaymentVaultType,
+                    salePrice: self.details.salePrice,
+                    customID: self.details.customID,
+                    commissionAmount: self.details.commissionAmount,
```

Comment:
I don't think it is needed as `purchased: self.details.purchased` will cover this as it will be false instead of true (if the real purchase happens). Anyway, all the details we are emitting are the same that are added during the creation of the listing so no need to change them.